ufl/core/external_operator.py

```diff
@@ -0,0 +1,98 @@
+from ufl.utils.str import as_native_str
+from ufl.coefficient import Coefficient
+from ufl.core.ufl_type import ufl_type
+from ufl.constantvalue import as_ufl
+from ufl.log import error
+from ufl.finiteelement.mixedelement import TensorElement
+from ufl.functionspace import FunctionSpace
+from ufl.domain import default_domain
+import weakref
+
+
+@ufl_type(inherit_indices_from_operand=0, is_terminal=True, is_differential=True)
+class ExternalOperator(Coefficient):
+
+    # Slots are disabled here because they cause trouble in PyDOLFIN
+    # multiple inheritance pattern:
+    # __slots__ = as_native_strings(("ufl_operands", "_ufl_function_space", "derivatives", "_ufl_shape"))
+    _ufl_noslots_ = True
+    _ufl_all_external_operators_ = weakref.WeakKeyDictionary()
+
+    def __init__(self, *operands, function_space, derivatives=None, count=None):
+
+        # Check operands
+        for i in operands:
+            as_ufl(i)
+
+        self.ufl_operands = operands
+
+        Coefficient.__init__(self, function_space, count)
+
+        # Checks
+        if derivatives is not None:
+            if not isinstance(derivatives, tuple):
+                error("Expecting a tuple for derivatives and not %s" % derivatives)
+            if len(derivatives) == len(self.ufl_operands):
+                self.derivatives = derivatives
+                s = self.ufl_shape
+                for i, e in enumerate(self.derivatives):
+                    s += self.ufl_operands[i].ufl_shape * e
+                if len(s) > len(self.ufl_shape):
+                    sub_element = self._ufl_function_space.ufl_element()
+                    if len(sub_element.sub_elements()) != 0:
+                        sub_element = sub_element.sub_elements()[0]
+                    ufl_element = TensorElement(sub_element, shape=s)
+                    domain = default_domain(ufl_element.cell())
+                    self.original_function_space = FunctionSpace(domain, ufl_element)
+                    self._ufl_shape = ufl_element.reference_value_shape()
+                else:
+                    self._ufl_shape = self._ufl_function_space.ufl_element().reference_value_shape()
+                    self.original_function_space = self._ufl_function_space
+            else:
+                error("Expecting a size of %s for %s" % (len(self.ufl_operands), derivatives))
+        else:
+            self._ufl_shape = self._ufl_function_space.ufl_element().reference_value_shape()
+            self.original_function_space = self._ufl_function_space
+            self.derivatives = (0,) * len(self.ufl_operands)
```

Comment:
This object appears to be doing a lot of work, with, for example, a special way of describing derivatives. It would be good to explain what is going on.

Reply:
Ok I will do it